File: comixed-frontend/src/app/library/services/library.service.ts
/*
 * ComiXed - A digital comic book library management application.
 * Copyright (C) 2019, The ComiXed Project
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses>
 */

import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { interpolate } from 'app/app.functions';
import { GetLibraryUpdatesRequest } from 'app/library/models/net/get-library-updates-request';
import { LoggerService } from '@angular-ru/logger';
import { Observable } from 'rxjs';
import {
  CLEAR_IMAGE_CACHE_URL,
  CONSOLIDATE_LIBRARY_URL,
  CONVERT_COMICS_URL,
  DELETE_MULTIPLE_COMICS_URL,
  GET_LIBRARY_UPDATES_URL,
  START_RESCAN_URL,
  UNDELETE_MULTIPLE_COMICS_URL
} from 'app/library/library.constants';
import { ConvertComicsRequest } from 'app/library/models/net/convert-comics-request';
import { Comic } from 'app/comics';
import { ConsolidateLibraryRequest } from 'app/library/models/net/consolidate-library-request';
import { UndeleteMultipleComicsRequest } from 'app/library/models/net/undelete-multiple-comics-request';

@Injectable({
  providedIn: 'root'
})
export class LibraryService {
  constructor(private http: HttpClient, private logger: LoggerService) {}

  getUpdatesSince(
    lastUpdateDate: Date,
    lastComicId: number,
    maximumComics: number,
    processingCount: number,
    timeout: number
  ): Observable<any> {
    this.logger.debug(
      `[POST] http request: getting comic updates: lastUpdateDate=${lastUpdateDate} lastComicId=${lastComicId} maximumComics=${maximumComics}`
    );
    return this.http.post(interpolate(GET_LIBRARY_UPDATES_URL), {
      lastUpdatedDate: lastUpdateDate.getTime(),
      lastComicId: lastComicId,
      maximumComics: maximumComics,
      processingCount: processingCount,
      timeout: timeout
    } as GetLibraryUpdatesRequest);
  }

  startRescan(): Observable<any> {
    this.logger.debug('[POST] http request: starting rescan');
    return this.http.post(interpolate(START_RESCAN_URL), { start: true });
  }

  deleteMultipleComics(ids: number[]): Observable<any> {
    this.logger.debug('[POST] http request: deleting multiple comics:', ids);
    const params = new HttpParams().set('comic_ids', ids.toString());
    return this.http.post(interpolate(DELETE_MULTIPLE_COMICS_URL), params);
  }

  undeleteMultipleComics(ids: number[]): Observable<any> {
    this.logger.debug('[POST] http request: undelete multiple comics:', ids);
    return this.http.post(interpolate(UNDELETE_MULTIPLE_COMICS_URL), {
      ids: ids
    } as UndeleteMultipleComicsRequest);
  }

  convertComics(
    comics: Comic[],
    archiveType: string,
    renamePages: boolean,
    deletePages: boolean,
    deleteOriginal: boolean
  ): Observable<any> {
    this.logger.debug(
      '[POST] http request: converting comics:',
      comics,
      archiveType,
      renamePages,
      deletePages,
      deleteOriginal
    );
    return this.http.post(interpolate(CONVERT_COMICS_URL), {
      ids: comics.map(comic => comic.id),
      archiveType: archiveType,
      renamePages: renamePages,
      deletePages: deletePages,
      deleteOriginal: deleteOriginal
    } as ConvertComicsRequest);
  }

  /**
   * Notifies the server to begin moving the entire library.
   *
   * @param deletePhysicalFiles the delete physical files flag
   * @param targetDirectory the target directory
   * @param renamingRule the renaming rules
   */
  moveComics(
    deletePhysicalFiles: boolean,
    targetDirectory: string,
    renamingRule: string
  ): Observable<any> {
    this.logger.debug('[POST] http request consolidate library:');
    this.logger.debug(`  deletePhysicalFiles=${deletePhysicalFiles}`);
    this.logger.debug(`  targetDirectory=${targetDirectory}`);
    this.logger.debug(`  renamingRule=${renamingRule}`);
    return this.http.post(interpolate(CONSOLIDATE_LIBRARY_URL), {
      deletePhysicalFiles: deletePhysicalFiles,
      targetDirectory: targetDirectory,
      renamingRule: renamingRule
    } as ConsolidateLibraryRequest);
  }

  clearImageCache(): Observable<any> {
    this.logger.debug('[DELETE] jhttp request: clear image cache');
    return this.http.delete(interpolate(CLEAR_IMAGE_CACHE_URL));
  }
}

File: comixed-services/src/main/java/org/comixed/service/library/LibraryService.java
/*
 * ComiXed - A digital comic book library management application.
 * Copyright (C) 2019, The ComiXed Project
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses>
 */

package org.comixed.service.library;

import java.io.File;
import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import lombok.extern.log4j.Log4j2;
import org.apache.commons.io.FileUtils;
import org.comixed.adaptors.ArchiveType;
import org.comixed.model.comic.Comic;
import org.comixed.model.tasks.TaskType;
import org.comixed.model.user.LastReadDate;
import org.comixed.repositories.comic.ComicRepository;
import org.comixed.service.task.TaskService;
import org.comixed.task.model.ConvertComicsWorkerTask;
import org.comixed.task.runner.Worker;
import org.springframework.beans.factory.ObjectFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@Log4j2
public class LibraryService {
  @Autowired private ComicRepository comicRepository;
  @Autowired private TaskService taskService;
  @Autowired private ReadingListService readingListService;
  @Autowired private ObjectFactory<ConvertComicsWorkerTask> convertComicsWorkerTaskObjectFactory;
  @Autowired private Worker worker;

  public List<Comic> getComicsUpdatedSince(
      String email, Date latestUpdatedDate, int maximumComics, long lastComicId) {
    this.log.debug(
        "Finding up to {} comics updated since {} for {}", maximumComics, latestUpdatedDate, email);

    List<Comic> comics =
        this.comicRepository.getLibraryUpdates(latestUpdatedDate, PageRequest.of(0, maximumComics));
    List<Comic> result = new ArrayList<>();

    final long timestamp = latestUpdatedDate.getTime();
    for (int index = 0; index < comics.size(); index++) {
      Comic comic = comics.get(index);
      final long thisTimestamp = comic.getDateLastUpdated().getTime();
      final Long thisId = comic.getId();
      this.log.debug(
          "Checking comic: timestamp: {} >= {} id: {} > {}",
          thisTimestamp,
          timestamp,
          thisId,
          lastComicId);
      boolean include = thisTimestamp > timestamp || thisId > lastComicId;

      if (include) {
        this.log.debug("Including comic: id={}", thisId);
        result.add(comic);
      }
    }

    this.log.debug("Loading reading lists");
    this.readingListService.getReadingListsForComics(email, result);

    this.log.debug("Returning {} updated comic{}", result.size(), result.size() == 1 ? "" : "s");
    return result;
  }

  public List<LastReadDate> getLastReadDatesSince(String email, Date lastReadDate) {
    return null;
  }

  public long getProcessingCount() {
    this.log.debug("Getting processing count");
    return this.taskService.getTaskCount(TaskType.PROCESS_COMIC);
  }

  public void convertComics(
      List<Long> comicIdList, ArchiveType targetArchiveType, boolean renamePages) {
    this.log.debug(
        "Converting {} comic{} to {}{}",
        comicIdList.size(),
        comicIdList.size() == 1 ? "" : "s",
        targetArchiveType,
        renamePages ? " (renaming pages)" : "");
    List<Comic> comics = new ArrayList<>();
    for (long id : comicIdList) {
      comics.add(this.comicRepository.getById(id));
    }
    this.log.debug("Getting save comics worker task");
    ConvertComicsWorkerTask task = this.convertComicsWorkerTaskObjectFactory.getObject();

    task.setComicList(comics);
    task.setTargetArchiveType(targetArchiveType);
    task.setRenamePages(renamePages);

    this.log.debug("Queueing save comics worker task");
    this.worker.addTasksToQueue(task);
  }

  @Transactional
  public List<Comic> consolidateLibrary(boolean deletePhysicalFiles) {
    this.log.debug("Consolidating library: delete physical files={}", deletePhysicalFiles);

    List<Comic> result = this.comicRepository.findAllMarkedForDeletion();

    for (Comic comic : result) {
      this.log.debug("Removing deleted comics from library");
      this.comicRepository.delete(comic);

      if (deletePhysicalFiles) {
        String filename = comic.getFilename();
        File file = comic.getFile();
        this.log.debug("Deleting physical file: {}", filename);
        FileUtils.deleteQuietly(file);
      }
    }
    return result;
  }
}
